src/raftstore/store/snap.rs

```diff
@@ -618,6 +620,987 @@ mod v1 {
+                                                   -> RaftStoreResult<Snap> {
+            let cf_size_checksums = try!(decode_cf_size_checksums(snapshot_data.get_data()));
+            let dir_path = dir.into();
+            if !dir_path.exists() {
+                try!(fs::create_dir_all(dir_path.as_path()));
+            }
+            let prefix = format!("{}_{}", SNAP_REV_PREFIX, key);
+            let display_path = Snap::get_display_path(&dir_path, &prefix);
+            debug!("receive a snapshot {}, with cf files sizes and checksums: {:?}",
+                   display_path,
+                   cf_size_checksums);
+            let mut cf_files = Vec::with_capacity(cf_size_checksums.len());
+            for (cf, size, checksum) in cf_size_checksums {
+                let filename = format!("{}_{}{}", prefix, cf, SST_FILE_SUFFIX);
+                let path = dir_path.join(filename).as_path().to_str().unwrap().to_owned();
+                let tmp_path = format!("{}{}", path, TMP_FILE_SUFFIX);
+                let cf_file = CfFile {
+                    cf: cf,
+                    path: path,
+                    size: size,
+                    tmp_path: tmp_path,
+                    checksum: checksum,
+                    ..Default::default()
+                };
+                cf_files.push(cf_file);
+            }
+            let meta_filename = format!("{}{}", prefix, META_FILE_SUFFIX);
+            let meta_path = dir_path.join(meta_filename).as_path().to_str().unwrap().to_owned();
+            let meta_tmp_path = format!("{}{}", meta_path, TMP_FILE_SUFFIX);
+            let meta_file = MetaFile {
+                data: snapshot_data,
+                path: meta_path,
+                tmp_path: meta_tmp_path,
+                ..Default::default()
+            };
+            let mut s = Snap {
+                display_path: display_path,
+                cf_files: cf_files,
+                cf_index: 0,
+                meta_file: meta_file,
+                size_track: size_track,
+            };
+            try!(s.init_for_receiving());
+            Ok(s)
+        }
+
+        pub fn new_for_applying<T: Into<PathBuf>>(dir: T,
+                                                  key: &SnapKey,
+                                                  size_track: Arc<RwLock<u64>>)
+                                                  -> RaftStoreResult<Snap> {
+            let mut s = try!(Snap::new(dir, key, size_track, false));
+            try!(s.init_for_applying());
+            Ok(s)
+        }
+
+        fn init_for_building(&mut self) -> io::Result<()> {
+            for cf_file in &mut self.cf_files {
+                // initialize sst file writer
+                let env_opt = EnvOptions::new();
+                let mut io_options = Options::new();
+                let per_level_compression = if cf_file.cf == CF_LOCK {
+                    parse_rocksdb_per_level_compression(NO_COMPRESSION_PER_LEVEL).unwrap()
```

Comment:
can we get the compression config for different CF from RocksDB directly? @zhangjinpeng1987